File: scripts/algorithms/M/Maximum Trailing Zeros in a Cornered Path/Maximum Trailing Zeros in a Cornered Path.java
// Runtime: 84 ms (Top 75.61%) | Memory: 82.50 MB (Top 26.83%)

class Solution {
    public int maxTrailingZeros(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] two = new int[m][n];
        int[][] five = new int[m][n];
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                two[i][j] = getFactorNum(grid[i][j], 2);
                five[i][j] = getFactorNum(grid[i][j], 5);
            }
        }
        
        int[][] left2 = new int[m][n];
        int[][] left5 = new int[m][n];
        for (int i = 0; i < m; ++i) {
            int sum2 = 0;
            int sum5 = 0;
            for (int j = 0; j < n; ++j) {
                sum2 += two[i][j];
                sum5 += five[i][j];
                left2[i][j] = sum2;
                left5[i][j] = sum5;
            }
        }
        
        int[][] right2 = new int[m][n];
        int[][] right5 = new int[m][n];
        for (int i = 0; i < m; ++i) {
            int sum2 = 0;
            int sum5 = 0;
            for (int j = n - 1; j >= 0; --j) {
                sum2 += two[i][j];
                sum5 += five[i][j];
                right2[i][j] = sum2;
                right5[i][j] = sum5;
            }
        }
        
        int[][] up2 = new int[m][n];
        int[][] up5 = new int[m][n];
        for (int j = 0; j < n; ++j) {
            int sum2 = 0;
            int sum5 = 0;
            for (int i = 0; i < m; ++i) {
                sum2 += two[i][j];
                sum5 += five[i][j];
                up2[i][j] = sum2;
                up5[i][j] = sum5;
            }
        }
        
        int[][] down2 = new int[m][n];
        int[][] down5 = new int[m][n];
        for (int j = 0; j < n; ++j) {
            int sum2 = 0;
            int sum5 = 0;
            for (int i = m - 1; i >= 0; --i) {
                sum2 += two[i][j];
                sum5 += five[i][j];
                down2[i][j] = sum2;
                down5[i][j] = sum5;
            }
        }
        
        int res = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                // left-up
                if (i != 0 || j != 0) {
                    int cur = Math.min(left2[i][j] + up2[i][j] - two[i][j], left5[i][j] + up5[i][j] - five[i][j]);
                    res = Math.max(res, cur);
                }
                
                // up-right
                if (i != 0 || j != n - 1) {
                    int cur = Math.min(right2[i][j] + up2[i][j] - two[i][j], right5[i][j] + up5[i][j] - five[i][j]);
                    res = Math.max(res, cur);
                }
                
                // right-down
                if (i != m - 1 || j != n - 1) {
                    int cur = Math.min(right2[i][j] + down2[i][j] - two[i][j], right5[i][j] + down5[i][j] - five[i][j]);
                    res = Math.max(res, cur);
                }
                
                // down-left
                if (i != m - 1 || j != 0) {
                    int cur = Math.min(left2[i][j] + down2[i][j] - two[i][j], left5[i][j] + down5[i][j] - five[i][j]);
                    res = Math.max(res, cur);
                }
            }
        }
        return res;
    }
    
    private int getFactorNum(int input, int factor) {
        int res = 0;
        while (input != 0 && input % factor == 0) {
            ++res;
            input /= factor;
        }
        return res;
    }
}

File: scripts/algorithms/M/Maximum Trailing Zeros in a Cornered Path/Maximum Trailing Zeros in a Cornered Path.cpp
// Runtime: 737 ms (Top 44.40%) | Memory: 107.9 MB (Top 75.00%)

array<int, 2> operator+(const array<int, 2> &l, const array<int, 2> &r) { return { l[0] + r[0], l[1] + r[1] }; }
array<int, 2> operator-(const array<int, 2> &l, const array<int, 2> &r) { return { l[0] - r[0], l[1] - r[1] }; }
int pairs(const array<int, 2> &p) { return min(p[0], p[1]); }

class Solution {
public:
int factors(int i, int f) {
    return i % f ? 0 : 1 + factors(i / f, f);
}
int maxTrailingZeros(vector<vector<int>>& grid) {
    int m = grid.size(), n = grid[0].size(), res = 0;
    vector<vector<array<int, 2>>> h(m, vector<array<int, 2>>(n + 1)), v(m + 1, vector<array<int, 2>>(n));
    for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j) {
            array<int, 2> f25 = { factors(grid[i][j], 2), factors(grid[i][j], 5) };
            v[i + 1][j] = v[i][j] + f25;
            h[i][j + 1] = h[i][j] + f25;
        }
    for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j) {
            auto v1 = v[i + 1][j], v2 = v[m][j] - v[i][j];
            auto h1 = h[i][j], h2 = h[i][n] - h[i][j + 1];
            res = max({res, pairs(v1 + h1), pairs(v1 + h2), pairs(v2 + h1), pairs(v2 + h2)});
        }
    return res;
}
};